File: dispatch/Python/dispatch_class.py
import time
import numba
from numba import int32, float32, float64
from numba.experimental import jitclass
from abc import ABC, abstractmethod

# Define the structure of each class with Numba-supported types
int_spec = [("value", int32)]
float_spec = [("value", float32)]
double_spec = [("value", float64)]

# Abstract base class (not needed for jitclass)
class Value(ABC):
    @abstractmethod
    def process(self, iterations: int):
        pass

    @abstractmethod
    def print(self):
        pass

# Concrete implementation for int values
@jitclass(int_spec)
class IntValue:
    def __init__(self, initial=42):
        self.value = initial

    def process(self, iterations: int):
        for _ in range(iterations):
            self.value += 10

    def print(self):
        print(f"Processed int: {self.value}")

# Concrete implementation for float values
@jitclass(float_spec)
class FloatValue:
    def __init__(self, initial=3.14):
        self.value = initial

    def process(self, iterations: int):
        for _ in range(iterations):
            self.value *= 2.0

    def print(self):
        print(f"Processed float: {self.value}")

# Concrete implementation for double values
@jitclass(double_spec)
class DoubleValue:
    def __init__(self, initial=2.718281828459045):
        self.value = initial

    def process(self, iterations: int):
        for _ in range(iterations):
            self.value /= 2.0

    def print(self):
        print(f"Processed double: {self.value}")

# Factory class to create appropriate value objects
class ValueFactory:
    @staticmethod
    def create_value(value_type: str):
        if value_type == "int":
            return IntValue()
        elif value_type == "float":
            return FloatValue()
        elif value_type == "double":
            return DoubleValue()
        else:
            raise ValueError(f"Unsupported type: {value_type}")

# Configuration reader class
class Config:
    @staticmethod
    def read_type(filename: str) -> str:
        with open(filename, 'r') as file:
            value_type = file.readline().strip()
        if not value_type:
            raise ValueError("Failed to read type from config file")
        return value_type

# Main execution
if __name__ == "__main__":
    try:
        # Read configuration
        config_file = "config.txt"
        value_type = Config.read_type(config_file)

        # Create value object based on configuration
        value = ValueFactory.create_value(value_type)

        # Start timing
        iterations = 1_000_000_000
        start_time = time.time()

        # Process the value multiple times
        value.process(iterations)

        # Stop timing
        elapsed_time = time.time() - start_time

        # Print results
        value.print()
        print(f"Time taken for {iterations} iterations: {elapsed_time:.6f} seconds")

    except Exception as e:
        print(f"Error: {e}")

File: dispatch/with_data/Rust/dispatch_class.rs
use std::fs;
use std::time::Instant;
use std::error::Error;
use std::fmt;

// Define a trait for values
trait Value: fmt::Display {
    fn process(&mut self);
}

// Implement the trait for different types
struct IntValue {
    value: i32,
}

impl IntValue {
    fn new(initial: i32) -> Self {
        IntValue { value: initial }
    }
}

impl Value for IntValue {
    fn process(&mut self) {
        self.value += 10; // Example: add 10 to int
    }
}

impl fmt::Display for IntValue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Processed int: {}", self.value)
    }
}

struct FloatValue {
    value: f32,
}

impl FloatValue {
    fn new(initial: f32) -> Self {
        FloatValue { value: initial }
    }
}

impl Value for FloatValue {
    fn process(&mut self) {
        self.value *= 2.0; // Example: multiply float by 2
    }
}

impl fmt::Display for FloatValue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Processed float: {}", self.value)
    }
}

struct DoubleValue {
    value: f64,
}

impl DoubleValue {
    fn new(initial: f64) -> Self {
        DoubleValue { value: initial }
    }
}

impl Value for DoubleValue {
    fn process(&mut self) {
        self.value /= 2.0; // Example: divide double by 2
    }
}

impl fmt::Display for DoubleValue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Processed double: {}", self.value)
    }
}

// Factory to create value objects
struct ValueFactory;

impl ValueFactory {
    fn create_value(type_name: &str) -> Result<Box<dyn Value>, Box<dyn Error>> {
        match type_name {
            "int" => Ok(Box::new(IntValue::new(42))),
            "float" => Ok(Box::new(FloatValue::new(3.14))),
            "double" => Ok(Box::new(DoubleValue::new(2.718281828459045))),
            _ => Err(format!("Unsupported type: {}", type_name).into()),
        }
    }
}

// Configuration reader
struct Config;

impl Config {
    fn read_type(filename: &str) -> Result<String, Box<dyn Error>> {
        let content = fs::read_to_string(filename)?;
        let type_name = content.trim().to_string();
        Ok(type_name)
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    // Read configuration
    let config_file = "config.txt";
    let type_name = Config::read_type(config_file)?;

    // Create value object based on configuration
    let mut value = ValueFactory::create_value(&type_name)?;

    // Start timing
    let iterations = 1_000_000_000;
    let start = Instant::now();

    // Process the value multiple times
    for _ in 0..iterations {
        value.process();
    }

    // Stop timing
    let elapsed = start.elapsed();

    // Print results
    println!("{}", value);
    println!(
        "Time taken for {} iterations: {:.2} seconds",
        iterations,
        elapsed.as_secs_f64()
    );

    Ok(())
}
